eclair-core/src/main/scala/fr/acinq/eclair/payment/send/OfferPayment.scala

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2023 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.payment.send
+
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.adapter.TypedActorRefOps
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
+import akka.actor.{ActorRef, typed}
+import fr.acinq.bitcoin.scalacompat.Crypto.PrivateKey
+import fr.acinq.bitcoin.scalacompat.{ByteVector32, ByteVector64}
+import fr.acinq.eclair.message.Postman.{OnionMessageResponse, SendMessage}
+import fr.acinq.eclair.message.{OnionMessages, Postman}
+import fr.acinq.eclair.payment.send.MultiPartPaymentLifecycle.PreimageReceived
+import fr.acinq.eclair.payment.send.PaymentInitiator.SendPaymentToNode
+import fr.acinq.eclair.payment.{Bolt12Invoice, PaymentFailed, PaymentSent}
+import fr.acinq.eclair.router.Router.RouteParams
+import fr.acinq.eclair.wire.protocol.OfferTypes.{InvoiceRequest, Offer}
+import fr.acinq.eclair.wire.protocol.{OfferTypes, OnionMessagePayloadTlv, TlvStream}
+import fr.acinq.eclair.{Features, InvoiceFeature, MilliSatoshi, NodeParams, TimestampSecond, randomKey}
+
+import java.util.UUID
+import scala.util.Random
+
+object OfferPayment {
+  sealed trait Result
+
+  sealed trait Failure extends Exception with Result
```

Comment:
We should avoid using `Exception` here, and only translate to an exception in `Eclair.scala` (at the boundaries with the HTTP API, where we don't have a choice because we need to use `Future.failed`).

Note that in the longer term, we don't really need to use `Future.failed` for the API, we could always return a `Future.successful` with an error object, which would be cleaner and better for performance. But that's for a very separate refactoring!

Reply:
Done